### PR TITLE
Fix #29: Document use of develop branch

### LINKED ISSUE #29: Cherry-picking Sucks
RDF4J project should have two active branches one targeting the the next minor release and the other targeting the next patch release. This will make patch release much less expensive and avoid potential human error involved in cherry-picking commits. Issues should be scheduled for a release prior to merging their PRs. Only once an PR is both approved *and* schedule should it be merged into the corresponding branch (i.e. develop for next minor release and master for next patch release). This pushes scheduling decisions to be made while the PR is open, reducing the confusion during the release process.

### PR DESCRIPTION
**Pull request by catch-point**
We should use a develop branch for features PRs and keep the master branch for patch releases.

You can read about it here: https://github.com/jamesrdf/rdf4j-doc/blob/issues/%2329-develop/doc/developer/index.adoc

Signed-off-by: James Leigh <james.leigh@ontotext.com>

### CODE REVIEW

**Review by abrokenjester — COMMENTED** [review]
Looks good, but I'd like to discuss on rdf4j-dev before we finalize this as "the new way to do things"

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: A switch from issues/#NNN-short-name to features/... will also have other references that need to be updated

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: I don't quite understand the logic for why the patch version is missing from master.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc:35]: If master has the same major/minor number as the latest release, that will imply in maven terms that it is prior to the last release, rather than after it. I would prefer to follow a principle of least surprise and increment the patch version on master after each release, rather than clearing the patch version and defaulting back to not having it.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: I don't quite understand what the reference to core and release archive is here. I would prefer that we define the public API and refer to it instead of adding novel terminology.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: "Once the Pull Request is created" looks like it belongs to point 2.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: This looks like it belongs to the point above.

Comment (ansell) [reply doc/developer/index.adoc]: Are unprivileged users able to update the milestone label or is that only for committers?

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: This looks like it belongs to the point above.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: This should only be done by a committer, given the sensitivity to merge conflicts. It isn't clear who the audience for this is right now.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: This needs a better description IMO. Something like "Pull Requests that add new features to the Public API"

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: This looks like it is part of the point above.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: This looks like it is part of the point above.

Comment (ansell) [reply doc/developer/index.adoc]: It isn't clear what "closing the branch" means in terms of users. The branch can happily stay in someones git repository without affecting anything, given that Git puts no special emphasis on "open" or "closed" branches.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: The only way to keep a Pull Request "open" is to not merge the changes. Otherwise GitHub will close it automatically. I am not sure I understand the rationale behind why the Pull Request state relates to the legal review, given legal reviews only apply to dependency changes (I thought).

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: What would the conditions for "not a release ready state" be? Is this just referring to the legal review for dependency updates?

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc:153]: This isn't optimal. Maven treats non-SNAPSHOT versions as if they are not going to change, and will confuse repositories who get conflicting versions from this branch. Best practice is to always use -SNAPSHOT on every commit except that which will be tagged as the release code.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc:159]: The backup files are ignored in .gitignore already, but this could still be useful.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc:167]: Using git tag -s to create a signed tag could be useful.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: I have never used "-u" with pushing of tags. Is it specifically useful?

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc:180]: I read the changes at the top to mean that this workflow would change from "git checkout 2.1.5" to "git checkout master"

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc:251]: Changing this to "git tag -s" to create a signed tag could be useful

**Review by ansell — CHANGES_REQUESTED** [review]
Clarification on a few points, including the use of SNAPSHOT versions consistently except on commits that are going to be tagged, the omission of the patch version from the master branch, and some clarification about the targeting and closing of pull requests

File: doc/developer/index.adoc
Comment (catch-point) [reply doc/developer/index.adoc]: This allows it to be easily pulled into the _develop_ branch (and reduce the amount of versioning commits in the history too).

File: doc/developer/index.adoc
Comment (catch-point) [reply doc/developer/index.adoc]: The feature branches are to group multiple issues together, such as if a new module needs collaboration while it is being created.

File: doc/developer/index.adoc
Comment (catch-point) [reply doc/developer/index.adoc]: This is here to indicate that even though a code change is beneficial and welcome, there may be other reason that require it to be delayed outside of the contributors control.

File: doc/developer/index.adoc
Comment (catch-point) [reply doc/developer/index.adoc]: Such as the code is ready, but the corresponding documentation is not ready.

File: doc/developer/index.adoc
Comment (catch-point) [reply doc/developer/index.adoc:167]: git tag -s only works if it has already been setup for that user

File: doc/developer/index.adoc
Comment (catch-point) [reply doc/developer/index.adoc:180]: Hotfixes are not based on master, they are based on a preceding release (not the latest). Like a bug fix release for java7 users.

Comment (abrokenjester) [reply doc/developer/index.adoc:180]: I think the confusion is that what you describe as a hotfix release is essentially a patch release for an _older_ minor/major release (whereas a patch release is for the _current_ major/minor).

**Review by abrokenjester — COMMENTED** [review]
I still agree with Peter that using a SNAPSHOT version is preferable until you’re ready to actually release, and am also not convinced that leaving out the patch version on master is a good idea (I get why you want it - it avoids conflicts when merging). 

All in all I’m not convinced that this new approach offers substantial benefits over what we currently do. The main goal of a change was to reduce the amount of cherry-picking needed, I feel we can do that with a far less radical change to the workflow, e.g. just making sure the patch release branch gets created earlier so that people can merge bug fixes into it directly if they want. 

But I don’t want to block things either. If you and the others want to give this a go, it’s up to you.

File: doc/developer/index.adoc
Comment (ansell) [reply doc/developer/index.adoc]: That sounds useful in some situations, but probably needs some more description here and promotion of the typical single-issue-branch case above this, as it isn't clear when to use issues/#NNN-issue-name and when to use features/feature-name with this coming first.

File: doc/developer/index.adoc
Comment (ansell) [reply doc/developer/index.adoc]: Thanks for the explanation, I hadn't been taking merge conflicts into account for that case. However, if it is rare to be releasing patch versions for non-current minor versions (then it may not be a big part of the workflow and may be confusing and not consistently followed by people familiar with how Maven generally works in other projects. But, as Jeen says, feel free to experiment and get some idea of whether it is less hassle.

File: doc/developer/index.adoc
Comment (ansell) [reply doc/developer/index.adoc:153]: Even if the release branch is done locally, unless you squash it down to a single commit before tagging and pushing, it would be useful to keep the -SNAPSHOT version until the actual commit that will be tagged. Squashing before tagging and pushing is definitely an option if the release manager is doing this all on their machine but would require some more instructions here.

File: doc/developer/index.adoc
Comment (ansell) [inline doc/developer/index.adoc]: It isn't clear who does the labelling. I.e., release manager, any committer, the contributor?

Comment (abrokenjester) [reply doc/developer/index.adoc]: It can't be a contributor as they do not have permission to edit the milestone field, AFAIK. So it would have to be committers or the release manager.

File: doc/developer/index.adoc
Comment (catch-point) [reply doc/developer/index.adoc]: Did my clarification (committed 5 days ago) help clear this up? Or, do you think more is needed?

File: doc/developer/index.adoc
Comment (catch-point) [reply doc/developer/index.adoc:153]: The idea is that the release branch will only contain one extra commit (the version change). All other changes should be done in master before starting the release prepping stage. What could be added here to clarify that?

**Comment by catch-point** [comment]
I think we need to move to a two active branch workflow. I also think we need to be able to easily move fixes from one active branch to another without dealing with conflicts. What I have outlined here is based on what has worked for other projects.

The two main changes are that the latest code would be in develop branch and the master version is not updated.

I think having experienced feature contributors use the develop branch for the latest code is actually a good thing as bugfix contributors would have an easier time aligning their production system with the master branch, since it will be very similar to the latest release.

What I outlined so far is to have a stale master version number that is not updated with each release. I have been suggesting omitting the patch number in master, but we could also use a high patch number to ensure maven knows that master is newer than any patch release. If we include a fixed patch number like 2.2.9-SNAPSHOT, were the 9 is always present, we would ensure maven treats it newer than any of the released patches. Peter, would this help with the workflow in your opinion?

**Comment by abrokenjester** [comment]
If it's a choice between no patch number and some fixed patch number that we recycle, I'd prefer to not use a patch number at all. None of this is "etched in stone" anyway. We can give your approach a try, and adapt if we encounter problems.

**Review by abrokenjester — APPROVED** [review]
I have some minor reservations but at this point I think we should just try it out, and see if/where it needs improvement. Assuming Peter feels his comments have been addressed I'm fine with merging this and implementing this workflow.

File: doc/developer/index.adoc
Comment (ansell) [reply doc/developer/index.adoc]: The first use of "features/" here is still ambiguous when reading through the document from top to bottom, as the specifics are described below this point, and "issues/#" is first introduced below here. It isn't clear why you need to introduce the "features/" convention, (which has a high likelihood of only being used a few times a year) here, given that the other 90% of PRs will be compact and use "issues/#" straight from develop or master. However, feel free to merge this and incrementally fix up bits like that.

**Review by ansell — APPROVED** [review]
Looks okay for an experimental run. It shifts a lot of complexity from release managers onto casual contributors, and deviates from the typical Maven conventions, but every project is unique.

**Pull request merged**
